skrub/_interpolation_join.py

```diff
@@ -0,0 +1,341 @@
+import joblib
+import pandas as pd
+from sklearn.base import BaseEstimator, TransformerMixin, clone
+from sklearn.compose import ColumnTransformer
+from sklearn.ensemble import (
+    HistGradientBoostingClassifier,
+    HistGradientBoostingRegressor,
+)
+from sklearn.utils._tags import _safe_tags
+
+from skrub._table_vectorizer import TableVectorizer
+
+
+class InterpolationJoin(TransformerMixin, BaseEstimator):
+    """Join with a table augmented by machine-learning predictions.
+
+    This is similar to a usual equi-join, but instead of looking for actual
+    rows in the right table that satisfy the join condition, we estimate what
+    those rows would contain if they existed in the table.
+
+    Suppose we want to join a table ``buildings(latitude, longitude, n_stories)``
+    with a table ``annual_avg_temp(latitude, longitude, avg_temp)``. Our annual
+    average temperature table may not contain data for the exact latitude and
+    longitude of our buildings. However, we can interpolate what we need from
+    the data points it does contain. Using ``annual_avg_temp``, we train a
+    model to predict the temperature, given the latitude and longitude. Then,
+    we use this model to estimate the values we want to add to our
+    ``buildings`` table. In a way we are joining ``buildings`` to a virtual
+    table, in which rows for any (latitude, longitude) location are inferred,
+    rather than retrieved, when requested. This is done with::
+
+        InterpolationJoin(annual_avg_temp, on=["latitude", "longitude"]).fit_transform(
+            buildings
+        )
+
+    Parameters
+    ----------
+    aux_table : DataFrame
+        The (auxiliary) table to be joined to the `main_table` (which is the
+        argument of ``transform``). ``aux_table`` is used to train a model that
+        takes as inputs the contents of the columns listed in ``aux_key``, and
+        predicts the contents of the other columns. In the example above, we
+        want our transformer to add temperature data to the table it is
+        operating on. Therefore, ``aux_table`` is the ``annual_avg_temp``
+        table.
+
+    main_key : list of str, or str
+        The columns in the main table used for joining. The main table is the
+        argument of ``transform``, to which we add information inferred using
+        ``aux_table``. The column names listed in ``main_key`` will provide the
+        inputs (features) of the interpolators at prediction (joining) time. In
+        the example above, ``main_key`` is ``["latitude", "longitude"]``, which
+        refer to columns in the ``buildings`` table. When joining on a single
+        column, we can pass its name rather than a list: ``"latitude"`` is
+        equivalent to ``["latitude"]``.
+
+    aux_key : list of str, or str
```

Comment:
Is this a necessary list, or could it be any iterable?

Reply:
Right, well spotted, thanks. 

For the sake of argument, since we already behave slightly differently than Pandas, should we be more flexible in terms of parameters and accept any iterable? This discussion is also useful for the `AggJoiner` PR.

Reply:
sure -- we do accept any iterable in practice (it gets converted to a [list](https://github.com/skrub-data/skrub/pull/742/files#diff-5988773cfa49ef9e6c03e0039fa0456a9df0d2d53ba6e38d13bb6bfee9572be4R227) ). so you mean change the docstring to say Iterable of str instead of list of str?